cosi/cosi.go

```diff
@@ -0,0 +1,502 @@
+	return &Response{Response: c.response}, err
+}
+
+// Response generates the response from the commitment, challenge and the
+// responses of its children.
+func (c *Cosi) Response(responses []*Response) (*Response, error) {
+	// //create your own response
+	if err := c.genResponse(); err != nil {
+		return nil, err
+	}
+	aggregateResponse := c.suite.Secret().Zero()
+	for _, resp := range responses {
+		// add responses of child
+		aggregateResponse = aggregateResponse.Add(aggregateResponse, resp.Response)
+		// add responses of it's children if there is one (i.e. if it is not a
+		// leaf)
+		if resp.ChildrenResp != nil {
+			aggregateResponse = aggregateResponse.Add(aggregateResponse, resp.ChildrenResp)
+		}
+	}
+	// Add our own
+	c.aggregateResponse = c.suite.Secret().Add(aggregateResponse, c.response)
+	return &Response{
+		Response:     c.response,
+		ChildrenResp: aggregateResponse,
+	}, nil
+}
+
+// Signature returns a signature in the like the EdDSA signature format plus the
+// bitmask at the end.
+// Sig = AggCommit || AggResponse || bitmask
+func (c *Cosi) Signature() []byte {
+	// Sig = R || S || bitmask
+	sigS := secretToSlice(c.aggregateResponse)
+	sigR, err := c.aggregateCommitment.MarshalBinary()
+	if err != nil {
+		panic("Can't generate signature !")
+	}
+	final := make([]byte, 64+c.mask.MaskLen())
+	copy(final[:], sigR)
+	copy(final[32:64], sigS)
+	copy(final[64:], c.mask.mask)
+	return final
+}
+
+// GetAggregateResponse returns the aggregated response that this cosi has
+// accumulated.
+func (c *Cosi) GetAggregateResponse() abstract.Secret {
+	return c.aggregateResponse
+}
+
+// GetChallenge returns the challenge that were passed down to this cosi.
+func (c *Cosi) GetChallenge() abstract.Secret {
+	return c.challenge
+}
+
+// GetCommitment returns the commitment generated by this CoSi (not aggregated).
+func (c *Cosi) GetCommitment() abstract.Point {
+	return c.commitment
+}
+
+// VerifyResponses verifies the response this CoSi has against the aggregated
+// public key the tree is using.
+// Reconstruct the AggCommit
+func (c *Cosi) VerifyResponses(aggregatedPublic abstract.Point) error {
+	k := c.challenge
+
+	// k * -aggPublic + s * B = k*-A + s*B
+	// from s = k * a + r => s * B = k * a * B + r * B <=> s*B = k*A + r*B
+	// <=> s*B + k*-A = r*B
+	minusPublic := c.suite.Point().Neg(aggregatedPublic)
+	kA := c.suite.Point().Mul(minusPublic, k)
+	sB := c.suite.Point().Mul(nil, c.aggregateResponse)
+	left := c.suite.Point().Add(kA, sB)
+
+	if !left.Equal(c.aggregateCommitment) {
+		return errors.New("recreated commitment is not equal to one given")
+	}
+
+	return nil
+}
+
+// VerifySignature is the method to call to verify a signature issued by a Cosi
+// struct. Publics is the WHOLE list of publics keys, the mask at the end of the
+// signature will take care of removing the indivual public keys that did not
+// participate
+func VerifySignature(suite abstract.Suite, publics []abstract.Point, message, sig []byte) error {
```

Comment:
I propose the following structure of the file:
- structures
- New-methods for the structures
- public methods on the structues
- private methods on the structures
- public library-methods
- private library-methods